homeassistant/components/irc.py

```diff
@@ -0,0 +1,303 @@
+                return
+
+            chan = self._channels[channel]
+            chan.users.clear()
+            for user in users.split(' '):
+                if user.startswith('@') or user.startswith('+'):
+                    user = user[1:]
+                chan.users.add(user)
+            chan.manual_update('users')
+
+        @irc3.event(irc3.rfc.JOIN_PART_QUIT)
+        def on_user_event(self, mask=None, event=None, channel=None, **kwargs):
+            """Handle JOIN, PART and QUIT commands."""
+            if channel not in self._channels:
+                return
+
+            nick = mask.split('!')[0]
+            chan = self._channels[channel]
+            if event == 'JOIN':
+                chan.users.add(nick)
+            elif nick in chan.users:
+                chan.users.remove(nick)
+            chan.manual_update('users')
+
+        @irc3.event(irc3.rfc.KICK)
+        def on_user_kick(self, channel, target, **kwargs):
+            """Handle KICK commands."""
+            if channel not in self._channels:
+                return
+
+            chan = self._channels[channel]
+            if target in chan.users:
+                chan.users.remove(target)
+            chan.manual_update('users')
+
+    return HAPlugin
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup the IRC component."""
+    hass.data[DATA_IRC] = {}
+
+    @asyncio.coroutine
+    def async_setup_irc_server(hass, config):
+        """Setup a new IRC server."""
+        irc_config = {
+            'loop': hass.loop,
+            'nick': config.get(CONF_NICK),
+            'host': config.get(CONF_HOST),
+            'port': config.get(CONF_PORT),
+            'username': config.get(CONF_USERNAME),
+            'realname': config.get(CONF_REAL_NAME),
+            'hass': hass,
+            'hass_config': config,
+            'autojoins': config.get(CONF_AUTOJOIN),
+            'ssl': config.get(CONF_SSL),
+            'includes': ['irc3.plugins.core', __name__]
+        }
+
+        if CONF_PASSWORD in config:
+            irc_config.update({'password': config.get(CONF_PASSWORD)})
+
+        if not config.get(CONF_VERIFY_SSL):
+            irc_config.update({'ssl_verify': 'CERT_NONE'})
+
+        import irc3
+        bot = irc3.IrcBot.from_config(irc_config)
+        network = config.get(CONF_NETWORK)
+        plugin_path = '{0}.{1}'.format(__name__, HAPlugin.__name__)
+        hass.data[DATA_IRC][network] = bot.get_plugin(plugin_path)
+        bot.run(forever=False)
+
+        hass.async_add_job(discovery.async_load_platform(
+            hass, 'notify', DOMAIN, {
+                CONF_NETWORK: network
+            }, config))
+
+    # Plugins must have an attribute which is not possible to add globally
+    # since Home Assistant installs dependencies on-the-fly. This adds it in
+    # runtime instead (sort of a hack).
+    global HAPlugin  # pylint: disable=invalid-name,global-variable-undefined
```

Comment:
Yeah, this is a fun little hack 😄 IRC3 uses venusian (https://pypi.python.org/pypi/venusian) to implement plugin support. It's supposed to make life easier for developer by only requiring decorators for certain things. To create a new plugin, decorate the class with irc3.plugin and add the module path to "includes" in the irc config (see async_setup_irc_server method). When a connection is setup to a server, it will scan all modules specified in "includes" and look for classes with that decorator and instantiate them, i.e. adding a new plugin.

This "global" hack defines a new class at top level of the module after it has been loaded